### PR TITLE
Implement bits of Module, ModuleFactory, ModuleData, ModuleInfo, ThingFactory and ThingTemplate

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/thing/moduleinfo.h [hunk 1]
```diff
@@ -19,30 +19,95 @@
...
+        Utf8String name;
+        Utf8String tag_name;
         ModuleData *data;
         int interface_mask;
         bool copied_from_default;
-        bool unk_bool1;
-        bool unk_bool2;
+        bool inherited;
+        bool like_kind_override;
+        Nugget(const Utf8String &n, const Utf8String &t, ModuleData *d, int m, bool i, bool l) :
```
Comment (tomsons26) [inline src/game/common/thing/moduleinfo.h]: bool inherited, bool like_kind

File: src/game/common/thing/thingfactory.cpp [hunk 2]
```diff
@@ -20,6 +20,10 @@
 ThingFactory *g_theThingFactory = nullptr;
 #endif
 
+#ifdef GAME_DEBUG_STRUCTS
+Utf8String &= Make_Global<Utf8String>(0x00E1D4F4);
```
Comment (tomsons26) [inline src/game/common/thing/thingfactory.cpp:24]: missing name

File: src/game/common/thing/thingtemplate.cpp [hunk 3]
```diff
@@ -13,10 +13,179 @@
...
+    {"EnergyProduction", &INI::Parse_Int, nullptr, offsetof(ThingTemplate, m_energyProduction)},
+    {"EnergyBonus", &INI::Parse_Int, nullptr, offsetof(ThingTemplate, m_energyBonus)},
+    {"IsForbidden", &INI::Parse_Bool, nullptr, offsetof(ThingTemplate, m_isForbidden)},
+    {"IsPrerequisite", &INI::Parse_Bool, nullptr, offsetof(ThingTemplate, m_isPrerequisite)},
+    {"DisplayColor", &INI::Parse_Color_Int, nullptr, offsetof(ThingTemplate, m_displayColor)},
+    {"EditorSorting", &INI::Parse_Byte_Sized_Index_List, s_editorSortingNames, offsetof(ThingTemplate, m_editorSorting)},
+    {"KindOf", &BitFlags<KINDOF_COUNT>::Parse_From_INI, nullptr, offsetof(ThingTemplate, m_kindOf)},
+    {"CommandSet", &INI::Parse_AsciiString, nullptr, offsetof(ThingTemplate, m_commandSetString)},
+    {"BuildVariations", &INI::Parse_AsciiString_Vector, nullptr, offsetof(ThingTemplate, m_buildVariations)},
+    {"Behavior", &ThingTemplate::Parse_Module_Name, nullptr, offsetof(ThingTemplate, m_body)},
```
Comment (tomsons26) [inline src/game/common/thing/thingtemplate.cpp]: (void *)0 for consistency

File: src/game/common/thing/thingtemplate.cpp [hunk 4]
```diff
@@ -297,8 +466,362 @@ void ThingTemplate::Resolve_Names()
...
+    m_draws.Clear();
+    m_clientUpdates.Clear();
+
+    Utf8String tag;
+    tag.Format("LTA_%sDestroyDie", m_LTAName.Str());
+    m_body.Add_Module_Info(this,
+        "DestroyDie",
+        tag,
+        g_theModuleFactory->New_Module_Data_From_INI(nullptr, "DestroyDie", MODULE_DEFAULT, tag),
+        2,
```
Comment (tomsons26) [inline src/game/common/thing/thingtemplate.cpp]: MODULEINTERFACE_DIE

File: src/game/common/thing/thingtemplate.cpp [hunk 5]
```diff
@@ -297,8 +466,362 @@ void ThingTemplate::Resolve_Names()
...
+        false,
+        false);
+
+    tag;
+    tag.Format("LTA_%sInactiveBody", m_LTAName.Str());
+    m_body.Add_Module_Info(this,
+        "InactiveBody",
+        tag,
+        g_theModuleFactory->New_Module_Data_From_INI(nullptr, "InactiveBody", MODULE_DEFAULT, tag),
+        32,
```
Comment (tomsons26) [inline src/game/common/thing/thingtemplate.cpp]: MODULEINTERFACE_BODY

File: src/game/common/thing/thingtemplate.cpp [hunk 6]
```diff
@@ -297,8 +466,362 @@ void ThingTemplate::Resolve_Names()
...
+        false,
+        false);
+
+    tag;
+    tag.Format("LTA_%sW3DDefaultDraw", m_LTAName.Str());
+    m_body.Add_Module_Info(this,
+        "W3DDefaultDraw",
+        tag,
+        g_theModuleFactory->New_Module_Data_From_INI(nullptr, "W3DDefaultDraw", MODULE_W3D, tag),
+        1024,
```
Comment (tomsons26) [inline src/game/common/thing/thingtemplate.cpp]: MODULEINTERFACE_DRAW

File: src/game/common/thing/moduleinfo.h (hunk 1 above)
Comment (jonwil) [reply src/game/common/thing/moduleinfo.h]: fixed.

File: src/game/common/thing/thingfactory.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/common/thing/thingfactory.cpp:24]: fixed.

File: src/game/common/thing/thingtemplate.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/common/thing/thingtemplate.cpp]: fixed.

File: src/game/common/thing/thingtemplate.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/common/thing/thingtemplate.cpp]: fixed.

File: src/game/common/thing/thingtemplate.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/common/thing/thingtemplate.cpp]: fixed.

File: src/game/common/thing/thingtemplate.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/common/thing/thingtemplate.cpp]: fixed.

File: src/game/common/thing/thingtemplate.cpp [hunk 7]
```diff
@@ -297,8 +466,364 @@ void ThingTemplate::Resolve_Names()
...
+        captainslog_relassert(tmp >= 0 && tmp < 65535, 0xDEAD0001, "Value parsed outside range of a short.");
+
+        *static_cast<uint16_t *>(store) = tmp;
+        tmplate->m_determinedBySuperweaponRestriction = false;
+    }
+}
+
+bool ThingTemplate::Remove_Module_Info(const Utf8String &tag, Utf8String &name)
+{
+    bool ret = m_body.Clear_Module_Data_With_Tag(tag, name);
```
Comment (tomsons26) [inline src/game/common/thing/thingtemplate.cpp]: lets keep it in a if like the rest for easier reading

Comment (jonwil) [reply src/game/common/thing/thingtemplate.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**